0005-snarkyjs-ecdsa-sha.md

```diff
@@ -148,31 +148,125 @@ Overall, exposing new gadgets and gates follow a strict pattern that has been us
+    ?(doubles : f Affine.t array option)
+    (signature :
+      f Foreign_field.Element.Standard.t * f Foreign_field.Element.Standard.t )
+    (msg_hash : f Foreign_field.Element.Standard.t)
+```
+
+The function takes the following arguments:
+
+- `base_check` := Context to track required base field external checks
+- `scalar_checks` := Context to track required scalar field external checks
+- `curve` := Elliptic curve parameters - for now, the goal is to make ECDSA over secp256k1 available. However, ECDSA accepts different curve parameters as well.
+- `pubkey` := Public key of signer
+- `doubles` := Optional powers of $2^i$ of the `pubkey`, $0 <= i < n$ where $n$ is `curve.order_bit_length`
+- `signature` := ECDSA signature (r, s) s.t. r, s $\in [1, n)$
+- `msg_hash` := Message hash s.t. msg_hash \in Fn - this will be the output of `Keccak`
+
+However, it is important to mention that the OCaml API has been designed with the assumption in mind that some of the inputs already satisfy a set of preconditions. These preconditions will not be checked internally, but are a requirement. Additionally, verifying ECDSA signatures is an expensive task and requires a lot of constraints. By moving the precondition checks of the inputs outside of the actual verification step (requiring them as preconditions), it allows us to optimize constraints but also provides a bigger API surface to cover in order for us to provide a safe API developers can use.
+
+The input preconditions are:
+
+- `pubkey` is on the curve and not O (`Ec_group.is_on_curve` gadget)
+- `pubkey` is in the subgroup (nP = O) (`Ec_group.check_subgroup` gadget)
+- `pubkey` is bounds checked (`multi-range-check` gadgets)
+- `r, s` $\in [1, n)$ (`signature_scalar_check` gadget)
+- `msg_hash` $\in Fn$ (`bytes_to_foreign_field_element` gadget)
+
+Each of these preconditions requires expensive checks so its important to choose wisely when and how to check these inputs. Its important to provide a safe API as well as giving experienced developers enough space to optimize their applications by avoiding double constraining of inputs.
+
+By building on top of the [`ForeignField`](https://github.com/o1-labs/snarkyjs/pull/985) implementation, we can leverage the modular approach taken by it to implement an ECDSA API which provides developers with a powerful modular interface. This approach allows us to potentially enable multiple versions of ECDSA with different curves, within the same API. We will also implement a `ForeignCurve` API which relies on the non-native elliptic curve primitives and gadgets introduce in the original [ECDAS PR](https://github.com/MinaProtocol/mina/pull/13279) so developers can not only profit from a modular ECDSA API, but can also access non-native elliptic curves directly.
+
+Similar to [`ForeignField`](https://github.com/o1-labs/snarkyjs/pull/985), the API of `ForeignCurve` will make use of a modular approach.
```

Comment:
this is all @mitschabaude - I just compiled it into the RFC. Thank you very much!